### ISSUE TITLE
 Exception occurred when getting the state dictionary（state_dict）.

### ISSUE THREAD

**Issue by lintao185** [dotnet/TorchSharp#1272]
![image](https://github.com/dotnet/TorchSharp/assets/55724885/87448f8c-6a05-44f2-a6d7-15a2a549c1e1)
![image](https://github.com/dotnet/TorchSharp/assets/55724885/83bdcdfa-be1b-4d12-91ee-abf6746a7e7c)
![image](https://github.com/dotnet/TorchSharp/assets/55724885/9f6c9fa3-3a8e-4bf9-bf3f-e2e3924d18c6)

In TorchSharp, I defined a model that contains nn.Sequential cv4. However, when I obtained the state_dict of the entire model, the dictionary for cv4 was missing, which is very strange. Other models also have nn.Sequential, and they can all be correctly obtained, but not the last layer.

**Comment by lintao185** [comment]
I suspect it's because of this API that it fails to recognize all the modules.
this.add_module(nameof(model), this.model);

**Comment by lintao185** [comment]
![image](https://github.com/dotnet/TorchSharp/assets/55724885/303ab0db-a139-4f96-bea6-3a24f3e0a18d)
![image](https://github.com/dotnet/TorchSharp/assets/55724885/2d94577f-a355-4547-9921-e3b780b9fa03)
When the elements of ModuleList are nn.Module<Tensor, Tensor> instead of nn.Sequential, the state_dict of the ModuleList elements can be captured correctly.

**Comment by lintao185** [comment]
When calling `model.add_module`, it traverses its sub-items and calls `RegisterComponents()` for each sub-item. However, `Sequential` does not perform such an operation. As a temporary measure, in a custom model, after initialization, force the call to `RegisterComponents()`.
![image](https://github.com/dotnet/TorchSharp/assets/55724885/63d48e4f-d593-447d-91eb-e29e22c9ca89)

**Comment by lintao185** [comment]
Comparing Sequential and ModuleList, it can be observed that ModuleList overrides RegisterComponents, whereas Sequential does not. Because ModuleList overrides RegisterComponents, it gains the ability to automatically invoke the RegisterComponents of its child items.
![image](https://github.com/dotnet/TorchSharp/assets/55724885/893d7135-f1df-4fef-b613-25ef5d5951f3)
![image](https://github.com/dotnet/TorchSharp/assets/55724885/fb0332a6-2959-45e1-84ae-392130f14ed2)

**Comment by NiklasGustafsson** [comment]
Is there a smallish repro case that I can debug?

**Comment by lintao185** [comment]
```
public class AModel : TorchSharp.torch.nn.Module
{
    public TorchSharp.Modules.Sequential cv1;
    public TorchSharp.Modules.ModuleList<torch.nn.Module<torch.Tensor, torch.Tensor>> cv2;
    public AModel():base(nameof(AModel))
    {
        cv1 = nn.Sequential(Enumerable.Range(0, 3).Select(x => new BModel()));
        cv2 = nn.ModuleList<torch.nn.Module<torch.Tensor, torch.Tensor>>(Enumerable.Range(0, 3).Select(x => new BModel()).ToArray());
        RegisterComponents();
    }
}

public class BModel : torch.nn.Module<torch.Tensor,torch.Tensor>
{
    public Tensor stride;
    public Tensor stride2;


    public BModel() : base(nameof(BModel))
    {
        stride = torch.ones(100, 100);
        stride2 = torch.ones(100, 100);
    }

    public override torch.Tensor forward(torch.Tensor input)
    {
        throw new NotImplementedException();
    }
}
```
```
var a = new AModel();
var aStateDict = a.state_dict();
```
In this case, the state_dict of cv1 cannot be obtained.
```
public class AModel : TorchSharp.torch.nn.Module
{
    public TorchSharp.Modules.Sequential cv1;
    public TorchSharp.Modules.ModuleList<torch.nn.Module<torch.Tensor, torch.Tensor>> cv2;
    public AModel() : base(nameof(AModel))
    {
        cv1 = nn.Sequential(Enumerable.Range(0, 3).Select(x => new BModel()));
        cv2 = nn.ModuleList<torch.nn.Module<torch.Tensor, torch.Tensor>>(Enumerable.Range(0, 3).Select(x => new BModel()).ToArray());
        RegisterComponents();
    }
}

public class BModel : torch.nn.Module<torch.Tensor, torch.Tensor>
{
    public Tensor stride;
    public Tensor stride2;


    public BModel() : base(nameof(BModel))
    {
        stride = torch.ones(100, 100);
        stride2 = torch.ones(100, 100);
        RegisterComponents();
    }

    public override torch.Tensor forward(torch.Tensor input)
    {
        throw new NotImplementedException();
    }
}
```
```
var a = new AModel();
var aStateDict = a.state_dict();
```
In this case, the state_dict of cv1 can be obtained.

**Comment by yueyinqiu** [comment]
So I guess the problem is that `Sequential.RegisterComponents` won't call `RegisterComponents` on its submodules.

**Comment by lintao185** [comment]
Sequential and ModuleList have different implementation methods for RegisterComponents.

**Comment by yueyinqiu** [comment]
I suppose the issue could be simply solved by adding a call to the submodule's `RegisterComponents` when adding it.

Actually in my opinion, all the modules should always call `RegisterComponents` themselves, so there is no need to deal with the submodules (because they will do that on their own) but it seems not because even `ModuleList` and `Sequential` is not doing that...

I'm a bit confused now... I can't even use:

```csharp
using static TorchSharp.torch.nn;

var l = ModuleList(Linear(1, 1));
Console.WriteLine(l.state_dict().Count); // 0
```

I believe this is ill designed...

**Comment by lintao185** [comment]
```
 protected override void RegisterComponents()
 {
     if (_registered) return;

     for (int i = 0; i < _list.Count; i++) {
         register_module($"{i}", _list[i]);
     }
     _registered = true;
 }
```
This is the implementation of ModuleList, and I think adding similar code in Sequential should be able to fix it.

**Comment by yueyinqiu** [comment]
> ```
>  protected override void RegisterComponents()
>  {
>      if (_registered) return;
> 
>      for (int i = 0; i < _list.Count; i++) {
>          register_module($"{i}", _list[i]);
>      }
>      _registered = true;
>  }
> ```
> 
> This is the implementation of ModuleList, and I think adding similar code in Sequential should be able to fix it.

However you will find that it's unable to use `var l = ModuleList(BModel()).state_dict();`, since `RegisterComponents` of `Linear` will not be called by itself, so your models' `RegisterComponents` is also not invoked. That's because `Linear` allows models to be dynamically appended, so we have to register them dynamically, instead of calling `RegisterComponents` only once.

So one solution might be to call the submodule's `RegisterComponents` in `Sequential.Add`. However it might make `RegisterComponents` be called too early, especially when the submodules are also mutable. So I'm not sure what the expected behavior should be.

And let me repeat my suggestion, perhaps all the modules should always call `RegisterComponents` themselves, and the parent modules should not care about the submodules.

**Comment by lintao185** [comment]
We can call RegisterComponents once in the top-level model, so that other models will be registered automatically, which is the most convenient.

**Comment by lintao185** [comment]
![image](https://github.com/dotnet/TorchSharp/assets/55724885/6a098664-5128-4ff1-91c5-5d2359086c23)
I wonder if this could serve as a relatively good solution.

**Comment by lintao185** [comment]
To facilitate the use of pre-trained weights in PyTorch, it is advisable to maintain consistency with PyTorch as much as possible.

**Comment by yueyinqiu** [comment]
> To facilitate the use of pre-trained weights in PyTorch, it is advisable to maintain consistency with PyTorch as much as possible.

That is what I mean. A module should register the parameters by themselves. In PyTorch it is done by `__setattr__`. However it's impossible for csharp. Perhaps we need a source generator? So complicated it could be :(

**Comment by lintao185** [comment]
I'm not sure if this is feasible or not, but the idea is to call RegisterComponents within the parameterless constructor of nn.Module. This way, when you create a custom model that inherits from nn.Module, it will automatically register itself.

**Comment by yueyinqiu** [comment]
> I'm not sure if this is feasible or not, but the idea is to call RegisterComponents within the parameterless constructor of nn.Module. This way, when you create a custom model that inherits from nn.Module, it will automatically register itself.

Unfortunately, that's impossible. The constructor of nn.Module is runned before the constructor of your model. Thus no values have been assigned to the properties and fields, so we can't register them...

**Comment by lintao185** [comment]
Switching gears, we could declare properties and then mark them with a custom attribute that has a “name” which will be used as the registration name. Subsequently, we could employ Fody to inject code into the getter and setter methods to handle the registration process. However, this approach is somewhat cumbersome.

**Comment by yueyinqiu** [comment]
Yes I suppose Fody/SourceGenerator could be a beautiful solution.

@NiklasGustafsson Could you please take a look at this?

**Comment by yueyinqiu** [comment]
I have made a demo here: [https://github.com/yueyinqiu/TorchSharp.AutoRegister](https://github.com/yueyinqiu/TorchSharp.AutoRegister)

**Comment by NiklasGustafsson** [comment]
> > I'm not sure if this is feasible or not, but the idea is to call RegisterComponents within the parameterless constructor of nn.Module. This way, when you create a custom model that inherits from nn.Module, it will automatically register itself.
> 
> Unfortunately, that's impossible. The constructor of nn.Module is runned before the constructor of your model. Thus no values have been assigned to the properties and fields, so we can't register them...

That is exactly right. That's why `RegisterComponents` exists and needs to be called last in the (custom) module constructor.

**Comment by NiklasGustafsson** [comment]
> Switching gears, we could declare properties and then mark them with a custom attribute that has a “name” which will be used as the registration name. Subsequently, we could employ Fody to inject code into the getter and setter methods to handle the registration process. However, this approach is somewhat cumbersome.

That capability already exists. For example, in the rewrite we're working on for some of the standard modules, which will enable more attributes to be exposed, the parameters of `Linear` are defined as:

```
            const string WeightComponentName = nameof(weight);
            const string BiasComponentName = nameof(bias);

            public Parameter? bias {
                get => _bias;
                set {
                    _bias?.Dispose();
                    _bias = value?.DetachFromDisposeScope() as Parameter;
                    ConditionallyRegisterParameter(BiasComponentName, _bias);
                }
            }

            public Parameter weight {
                get => _weight!;
                set {
                    if (value is null) throw new ArgumentNullException(nameof(weight));
                    if (value.Handle != _weight?.Handle) {
                        _weight?.Dispose();
                        _weight = (value.DetachFromDisposeScope() as Parameter)!;
                        ConditionallyRegisterParameter(WeightComponentName, _weight);
                    }
                }
            }

            [ComponentName(Name = BiasComponentName)]
            private Parameter? _bias;
            [ComponentName(Name = WeightComponentName)]
            private Parameter? _weight;
```

**Comment by NiklasGustafsson** [comment]
>> Perhaps all the modules should always RegisterComponents themselves, and the parent modules should not care about registering the sub-submodules of its submodules (that is, not to call RegisterComponents on its submodules).

Yes, this is exactly the intended protocol, and the documentation says so: 
https://github.com/dotnet/TorchSharp/wiki/Creating-Your-Own-TorchSharp-Modules

The exception from this rule will be modules (such as the `Linear` module shown above, where the parameters may be altered, and the property setter needs to conditionally register a component, which allows you to assign it 'null' as well as overwrite an already existing parameter.

**Comment by NiklasGustafsson** [comment]
> I have made a simplified demo here: https://github.com/yueyinqiu/TorchSharp.AutoRegister
> 
> PS: It's still impossible to get rid of the traditional constructors (and use a primary constructor instead), because we have to access the generated property. So sad :(

As much as I dislike relying on reflection, which the current scheme does (I dislike it because it prevents AOT), having to use source code generation adds complexity and something that has to be automated. That would be a last resort, I think.

The current scheme works fairly well as long as you follow the instructions very closely and don't so advanced stuff like the `Linear` module above. I don't know why you would allow setting the parameters after the module has been constructed, but PyTorch does, so TorchSharp should, too.

**Comment by lintao185** [comment]
So, to address the issue of Sequential not registering its sub-modules, for the time being, should we also rewrite Sequential’s RegisterComponents, just like we did with ModuleList?

**Comment by yueyinqiu** [comment]
> > > Perhaps all the modules should always RegisterComponents themselves, and the parent modules should not care about registering the sub-submodules of its submodules (that is, not to call RegisterComponents on its submodules).
> 
> Yes, this is exactly the intended protocol, and the documentation says so: https://github.com/dotnet/TorchSharp/wiki/Creating-Your-Own-TorchSharp-Modules
> 
> The exception from this rule will be modules (such as the `Linear` module shown above, where the parameters may be altered, and the property setter needs to conditionally register a component, which allows you to assign it 'null' as well as overwrite an already existing parameter.

So it is my understanding that custom modules should not relies on others calling its `RegisterComponents`. But why we are doing that in `register_module`?

**Comment by NiklasGustafsson** [comment]
> So, to address the issue of Sequential not registering its sub-modules, for the time being, should we also rewrite Sequential’s RegisterComponents, just like we did with ModuleList?

We can certainly reconsider the protocol for module registration for the future. However, if the guidelines for custom modules described in the Wiki article are followed, the current protocol works.

**Comment by yueyinqiu** [comment]
> > > [...] why we are doing that in `register_module`?
> > 
> > 
> > As unsatisfying as this answer is -- I don't recall.
> 
> haha... Perhaps we should check all the modules provided by TorchSharp, and remove this call if nothing depends on that? I believe this should be done as early as possible, to avoid more projects' relying on it by mistake.

oh well there is at least one thing (`ModuleList`) that depends on that:

```csharp
using static TorchSharp.torch.nn;

var l = ModuleList(Linear(1, 1));
Console.WriteLine(l.state_dict().Count); // 0
```

**Comment by lintao185** [comment]
> > So, to address the issue of Sequential not registering its sub-modules, for the time being, should we also rewrite Sequential’s RegisterComponents, just like we did with ModuleList?
> 
> We can certainly reconsider the protocol for module registration for the future. However, if the guidelines for custom modules described in the Wiki article are followed, the current protocol works.

Now it’s calling RegisterComponents() in the custom module, it just feels a bit verbose, haha.

**Comment by alinpahontu2912** [comment]
Hey @lintao185, I think this is related to https://github.com/dotnet/TorchSharp/issues/1426, which should now work. A release hasn't been made yet with this change included, but you can try building the nuggets yourself and see it if it helps. If not, I suggest following the instructions from here: https://github.com/dotnet/TorchSharp/wiki/Creating-Your-Own-TorchSharp-Modules
